### PR TITLE
CDAP-16674: k8s service registration retry logic with backoff

### PR DESCRIPTION
**Pull request by domnicjp**

### CODE REVIEW

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java [hunk 1]
```diff
@@ -266,15 +268,42 @@ private boolean createV1Service(CoreV1Api api, String serviceName, Discoverable
...
-      api.createNamespacedService(namespace, service, null, null, null);
-      LOG.info("Service created in kubernetes with name {} and port {}", serviceName, port.getPort());
-    } catch (ApiException e) {
-      // It means the service already exists. In this case we update the port if it is not the same.
-      if (e.getCode() == HttpURLConnection.HTTP_CONFLICT) {
-        return false;
+    // Maximum elapse-able time currently set to 10 minutes
+    ExponentialBackOff backoffGenerator
+        = new ExponentialBackOff.Builder().setMaxElapsedTimeMillis(600000).build();
+    boolean serviceCreatedSuccessfully = false;
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java]: just call it `serviceCreated`

**Comment by chtyim** [comment]
This doesn't resolve the problem as described in the JIRA. Because after max retry time, if it still fail, the process would still stuck. We should fix the process stuck problem first, regardless of having retry or not.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java [hunk 2]
```diff
@@ -266,15 +268,42 @@ private boolean createV1Service(CoreV1Api api, String serviceName, Discoverable
...
+    while (!serviceCreatedSuccessfully) {
+      try {
+        api.createNamespacedService(namespace, service, null, null, null);
+        LOG.info("Service created in kubernetes with name {} and port {}", serviceName, port.getPort());
+        serviceCreatedSuccessfully = true;
+      } catch (ApiException e) {
+        // It means the service already exists. In this case we update the port if it is not the same.
+        if (e.getCode() == HttpURLConnection.HTTP_CONFLICT) {
+          return false;
+        } else {
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java]: No need to have `else` have an `if` block returned

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java [hunk 3]
```diff
@@ -266,15 +268,42 @@ private boolean createV1Service(CoreV1Api api, String serviceName, Discoverable
...
+          if (e.getCause() instanceof SocketTimeoutException) {
+            LOG.warn("Service creation in kubernetes with name {} and port {} encountered exception: {}",
+                     serviceName,
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java]: No need to have that many new lines. Just have the arguments on the same line.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java [hunk 4]
```diff
@@ -266,15 +268,42 @@ private boolean createV1Service(CoreV1Api api, String serviceName, Discoverable
...
+                     port.getPort(),
+                     e);
+            try {
+              Thread.sleep(backoffDuration);
+              backoffDuration = backoffGenerator.nextBackOffMillis();
+            } catch (Exception f) {
+              LOG.warn("Encountered exception during k8s service creation backoff: " + f);
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java]: What kind of exception would it be? Is it interrupted exception? If you ignore this, it means the caller may be blocked inside this method and not able to cancel it (e.g. upon an explicit shutdown call).

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java [hunk 5]
```diff
@@ -266,15 +268,42 @@ private boolean createV1Service(CoreV1Api api, String serviceName, Discoverable
...
+        } else {
+          if (e.getCause() instanceof SocketTimeoutException) {
+            LOG.warn("Service creation in kubernetes with name {} and port {} encountered exception: {}",
+                     serviceName,
+                     port.getPort(),
+                     e);
+            try {
+              Thread.sleep(backoffDuration);
+              backoffDuration = backoffGenerator.nextBackOffMillis();
+            } catch (Exception f) {
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java]: don't use `f` as exception variable. Use something that starts with `e` or `ex`

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java [hunk 6]
```diff
@@ -266,15 +268,42 @@ private boolean createV1Service(CoreV1Api api, String serviceName, Discoverable
...
+        if (e.getCode() == HttpURLConnection.HTTP_CONFLICT) {
+          return false;
+        } else {
+          if (e.getCause() instanceof SocketTimeoutException) {
+            LOG.warn("Service creation in kubernetes with name {} and port {} encountered exception: {}",
+                     serviceName,
+                     port.getPort(),
+                     e);
+            try {
+              Thread.sleep(backoffDuration);
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java]: I am a bit concerned that we block the caller thread. Typically service registration shouldn't be blocked like this. It should either fail fast (like the current behavior), or retry in the background (which is kind of more complicated to handle).

**Comment by domnicjp** [comment]
Sorry for the delay in circling back on this.

> This doesn't resolve the problem as described in the JIRA.

Many apologies for misunderstanding the intent [there](https://issues.cask.co/browse/CDAP-16674?focusedCommentId=36915&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-36915) - in reading it I (certainly incorrectly on my part) thought of it is an either-or. Having said that, what follows are a few updates/notes/observations in that regard:

In testing with a modified image (that always errors in the `create` [and separately also in the `update`] code path), I see repetitive pod restarts (at least as I understand it - pardon my ignorance as it relates to all things `k8s`) and the eventual `CrashLoopBackOff` while steadily increasing `RESTART` counts (for the relevant `appfabric` pod). Here is an example from a recent test (with an image that as mentioned earlier always throws an exception in the `create`/`register` codepath + using a custom sandbox CLH to use this image for instance creation):

```
NAME                                                     READY   STATUS             RESTARTS   AGE
cdap-controller-0                                        1/1     Running            0          33m
cdap-ctd002-ap-006-appfabric-0                           0/1     CrashLoopBackOff   10         33m
cdap-ctd002-ap-006-logs-0                                1/1     Running            0          33m
...
```

...which seems to suggest restart attempts being made - should exceptions be thrown during the `create`/'service registration phase' of the (`appfabric`) pod initialization. So I am not entirely sure what, if anything ,needs to change to ensure that 'pod restart happens' as the JIRA observes.

I have attached [logs](https://issues.cask.co/secure/attachment/17022/af_create_02.log) and [thread dumps](https://issues.cask.co/secure/attachment/17025/af_create_thread_dump_03.out) to the ticket should that be handy in determining how best to proceed with the 'non-retry logic aspects' of fixing this here issue. To that end, [this patch](https://gist.github.com/domnicjp/ef1c628572e3adb9e75771c7493b7370) _may_ be of interest - which was born out of the observation of [this stacktrace](https://gist.github.com/domnicjp/fd3782e86ce5341d756c7a8765cf1ae9). However as I am not fully certain if that would not introduce other potential 'races'/side-effects, I have left it out for now.

> I am a bit concerned that we block the caller thread.
> Typically service registration shouldn't be blocked
> like this. It should either fail fast (like the current behavior),
> or retry in the background (which is kind of more
> complicated to handle).

With a view to simplifying the attempted retries, I have taken out the notions of the 'backoff' between retries (which eliminates the sleeps and concomitant blocked thread/'swallowed exceptions'/uninterruptible behaviors et al.). Hopefully this approach is better suited in this context - since as mentioned above the pod restarts _seem to be occurring_, so all we need to do is perhaps retry service registration to get past any transient latencies/`SocketTimeoutException`'s.

On a related note, it is quite possible that I am 'missing the forest for the trees' or likely not spotting the 'right details to focus on' in the thread dumps(for `create` and `update`), logs (for both). So please do let me know your thoughts on how best to address fixing this issue. Thanks.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java [hunk 7]
```diff
@@ -266,15 +266,24 @@ private boolean createV1Service(CoreV1Api api, String serviceName, Discoverable
...
-      api.createNamespacedService(namespace, service, null, null, null);
-      LOG.info("Service created in kubernetes with name {} and port {}", serviceName, port.getPort());
-    } catch (ApiException e) {
-      // It means the service already exists. In this case we update the port if it is not the same.
-      if (e.getCode() == HttpURLConnection.HTTP_CONFLICT) {
-        return false;
+    boolean serviceCreated = false;
+    int maxAttempts = 3;
+
+    for (int i = 1; i <= maxAttempts && !serviceCreated; ++i) {
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java:272]: I still think we shouldn't retry in the discovery service. If any form of retries is needed, it should be from the caller. However, I think it is sufficient that on service registration failure, we make sure the main process gets terminated and let K8s to restart the pod.

Comment (domnicjp) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java:272]: > I still think we shouldn't retry in the discovery service. If any form of retries is needed, it should be from the caller.

Sounds good. The other context that seemed relevant is in the '[service start](https://gist.github.com/domnicjp/949adcaecdd84a528f6e89c941108205)' - not sure if that is worthwhile pursuing.

> However, I think it is sufficient that on service registration failure, we make sure the main process gets terminated and let K8s to restart the pod.

As noted [above](https://github.com/cdapio/cdap/pull/12443#issuecomment-665851976) - the tests that I performed for intentional errors(or rather intentional exceptions thrown) in both the `create`/'service registration' and `update` codepaths seemed to suggest repeated retries to start the `appfabric` pod i.e. I do not see a 'wedged state' wherein the 'main process' remains 'un-terminated' or is preventing `k8s` from attempting to restart the pod.

Let me know your thoughts. (And thanks for your help with this!)

**Comment by domnicjp** [comment]
Per discussions here and [here](https://issues.cask.co/browse/CDAP-16674), closing this PR as "Won't Fix".

**Pull request closed without merging**